frame/fast-unstake/src/lib.rs

```diff
@@ -412,61 +454,71 @@ pub mod pallet {
 
 				let remaining = T::DepositCurrency::unreserve(&stash, deposit);
 				if !remaining.is_zero() {
-					frame_support::defensive!("`not enough balance to unreserve`");
-					ErasToCheckPerBlock::<T>::put(0);
-					Self::deposit_event(Event::<T>::InternalError)
+					Self::halt("not enough balance to unreserve");
 				} else {
 					log!(info, "unstaked {:?}, outcome: {:?}", stash, result);
 					Self::deposit_event(Event::<T>::Unstaked { stash, result });
 				}
+			};
 
-				<T as Config>::WeightInfo::on_idle_unstake()
-			} else {
-				// eras remaining to be checked.
-				let mut eras_checked = 0u32;
+			let check_stash = |stash, deposit, eras_checked: &mut u32| {
 				let is_exposed = unchecked_eras_to_check.iter().any(|e| {
 					eras_checked.saturating_inc();
 					Self::is_exposed_in_era(&stash, e)
 				});
 
-				log!(
-					debug,
-					"checked {:?} eras, exposed? {}, (v: {:?}, u: {:?})",
-					eras_checked,
-					is_exposed,
-					validator_count,
-					unchecked_eras_to_check.len()
-				);
-
-				// NOTE: you can be extremely unlucky and get slashed here: You are not exposed in
-				// the last 28 eras, have registered yourself to be unstaked, midway being checked,
-				// you are exposed.
 				if is_exposed {
 					T::DepositCurrency::slash_reserved(&stash, deposit);
 					log!(info, "slashed {:?} by {:?}", stash, deposit);
 					Self::deposit_event(Event::<T>::Slashed { stash, amount: deposit });
+					false
 				} else {
-					// Not exposed in these eras.
-					match checked.try_extend(unchecked_eras_to_check.clone().into_iter()) {
-						Ok(_) => {
-							Head::<T>::put(UnstakeRequest {
-								stash: stash.clone(),
-								checked,
-								deposit,
-							});
+					true
+				}
+			};
+
+			if unchecked_eras_to_check.is_empty() {
+				// `stash` is not exposed in any era now -- we can let go of them now.
+				stashes.into_iter().for_each(|(stash, deposit)| unstake_stash(stash, deposit));
+				Self::deposit_event(Event::<T>::Terminated);
+				<T as Config>::WeightInfo::on_idle_unstake()
+			} else {
+				// eras checked so far.
+				let mut eras_checked = 0u32;
+
+				let pre_length = stashes.len();
+				let stashes: BoundedVec<(T::AccountId, BalanceOf<T>), T::BatchSize> = stashes
+					.into_iter()
+					.filter(|(stash, deposit)| {
+						check_stash(stash.clone(), *deposit, &mut eras_checked)
+					})
+					.collect::<Vec<_>>()
+					.try_into()
+					.expect("filter can only lesson the length; still in bound; qed");
```

Comment:
I think these proofs are fine if the proven statement is directly above.  
Here it is 90 lines above, so someone who changes the code could miss it.  
I think we could use `truncate_from` here or invent some `defensive_truncate_from`.

Reply:
I actually wanted to make `defensive_truncate_from` but it has to be an extension trait living in `frame_support`, because bounded types are in `sp_core`.

But, fwiw, the filter is still 4 lines above :D I think you are confusing it with another statement over `stashes`.

Reply:
lesson -> lessen or better minimize.